book/src/asm/functions.md

```diff
@@ -0,0 +1,74 @@
+# Functions
+
+Machine functions are the entry points to this machine. They can be called from another machine or from the outside.
+
+For static machines, functions simply indicate which columns should be exposed as inputs and outputs. They do not contain statements, since the business logic of all functions in the static machine is defined directly in constraints. We refer to the previous example [here](./machines.md#static-machines).
+
+From now on, we describe functions in dynamic machines with the example of this simple machine:
+
+```
+{{#include ../../../test_data/asm/book/function.asm:all}}
+```
+
+## Function inputs and outputs
+
+> For dynamic machines, function inputs and outputs are not supported yet
+
+## Expressions
+
+### Field element literals
+
+Field element literals are signed elements of the prime field.
+
+```
+{{#include ../../../test_data/asm/book/function.asm:literals}}
+```
+
+### Registers
+
+Registers can be used as expressions, with the exception of assignment registers.
+```
+{{#include ../../../test_data/asm/book/function.asm:read_register}}
+```
+
+### Instructions
+
+Instructions which return outputs can be used as expressions.
+```
+{{#include ../../../test_data/asm/book/function.asm:instruction}}
+```
+
+## Statements
+
+### Labels
+
+Labels allow referring to a location in a function by name.
+
+```
+{{#include ../../../test_data/asm/book/function.asm:label}}
+```
+
+### Assignments
+
+Assignments allow setting the value of a write register to the value of an expression using an assignment register.
+
+```
+{{#include ../../../test_data/asm/book/function.asm:instruction}}
+```
+
+One important current requirement is for the assignment register of the assignment to be compatible with that of the expression. This is especially relevant for instructions: the assignment register of the instruction output must match that of the assignment. In this example, we use `Y` in the assignment as the output of `square` is `Y`:
+
+```
+{{#include ../../../test_data/asm/book/function.asm:square}}
+```
+
+
+> Specifying the assignment register of the assignment is currently required even in cases where it could be inferred.
```

Comment:
```suggestion
> Specifying the assignment register of the assignment is currently required even in cases where it could be inferred. That restriction may be lifted in the future.
```